### ISSUE TITLE
what's the use of gain ?

### ISSUE THREAD

**Issue by clearlovewl** [NVlabs/stylegan2-ada-pytorch#95]
why we need different gain after activation function and filter layer? 
i notice that gain of lrelu = sqrt(2) while linear = 1; gain of upsample layer=4 while other not.
what's the principle？

**Comment by zwishenzug** [comment]
I think it's related to the weight initialization.  If you look at StyleGAN 1, it passes this gain to the get_weight function.

he_std = gain / np.sqrt(fan_in)

Unlike regular he_std initialization, in StyleGAN, this term is used as a runtime scaling value (as part of the Equalized Learning Rate).

By StyleGAN 2, it seems '1' is always passed in for the gain during weight initialization, and instead the gain is applied after the activation.

I haven't really thought it through in any depth, but I'm guessing it's equivalent to do it after the activation instead of as part of the runtime coefficient and makes the code cleaner.

**Comment by clearlovewl** [comment]
thank you. I noticed network use some weighted convolution . it seems that it's a trick to control the backend grad flow.&nbsp;


 
---Original---
From: ***@***.***&gt;
Date: 2021/6/22 21:58:40
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [NVlabs/stylegan2-ada-pytorch] what's the use of gain ? (#95)




 
I think it's related to the weight initialization.  If you look at StyleGAN 1, it passes this gain to the get_weight function.
 
he_std = gain / np.sqrt(fan_in)
 
Unlike regular he_std initialization, in StyleGAN, this term is used as a runtime scaling value (as part of the Equalized Learning Rate).
 
By StyleGAN 2, it seems '1' is always passed in for the gain during weight initialization, and instead the gain is applied after the activation.
 
I haven't really thought it through in any depth, but I'm guessing it's equivalent to do it after the activation instead of as part of the runtime coefficient and makes the code cleaner.

**Comment by zwishenzug** [comment]
I think this part in the StyleGAN 2 paper refers to this change.

"In practice our design can be implemented efficiently using grouped convolutions, as detailed in Appendix B. To avoid having to account for the activation function in Equation 3, we scale our activation functions so that they retain the expected signal variance."
